source/trace_event.c

```diff
@@ -0,0 +1,191 @@
+    do {                                                                                                               \
+        if (!result) {                                                                                                 \
+            AWS_FATAL_ASSERT(0 && "trace failure")                                                                     \
+        }                                                                                                              \
+    } while (0)
+
+struct aws_trace_event {
+    struct cJSON *root, *event_array;
+    struct aws_bus bus;
+    struct aws_allocator *allocator;
+};
+
+/* Add trace event meta data to JSON object when the message bus allows it */
+void aws_trace_event_listener(uint64_t address, const void *msg, void *user_data) {
+    struct aws_trace_event_metadata *trace_event_data = (struct aws_trace_event_metadata *)msg;
+    if (trace_event_data == NULL) {
+        aws_raise_error(AWS_ERROR_OOM);
+    }
+    cJSON *event = cJSON_CreateObject();
+    if (event == NULL) {
+        aws_raise_error(AWS_ERROR_OOM);
+    }
+
+    cJSON_AddItemToArray(trace_event->event_array, event);
+
+    /* add options for args later on, will initially be empty for now */
+    cJSON_AddItemToObject(event, "args", cJSON_CreateObject());
+
+    /* add more options later on */
+    if (cJSON_AddStringToObject(event, "cat", trace_event_data->category) == NULL) {
+        aws_raise_error(AWS_ERROR_OOM);
+    }
+    if (cJSON_AddStringToObject(event, "name", trace_event_data->name) == NULL) {
+        aws_raise_error(AWS_ERROR_OOM);
+    }
+    /* Fix for buffer overflow issue using char phase reference */
+    char ph[2];
+    strncat(ph, &(trace_event_data->phase), 1);
+    if (cJSON_AddStringToObject(event, "ph", ph) == NULL) {
+        aws_raise_error(AWS_ERROR_OOM);
+    }
+    if (cJSON_AddNumberToObject(event, "pid", trace_event_data->process_id) == NULL) {
+        aws_raise_error(AWS_ERROR_OOM);
+    }
+    if (cJSON_AddNumberToObject(event, "tid", trace_event_data->thread_id)) {
+        aws_raise_error(AWS_ERROR_OOM);
+    }
+    if (cJSON_AddNumberToObject(event, "ts", (double)trace_event_data->timestamp)) {
+        aws_raise_error(AWS_ERROR_OOM);
+    }
+}
+
+/* Destructor of trace event metadata sent through the bus */
+void aws_trace_event_destroy(void *payload) {}
+
+void aws_trace_system_write(int time_unit, const char *filename) {
+    if (time_unit) {
+        if (cJSON_AddStringToObject(trace_event->root, "displayTimeUnit", "ns") == NULL) {
+            aws_raise_error(AWS_ERROR_OOM);
+        }
+    }
+    char *out = cJSON_Print((trace_event->root));
+    if (out == NULL) {
+        aws_raise_error(AWS_ERROR_OOM);
+    }
+    FILE *fp;
+    char fn[strlen(filename) + 6];
+
+    strcpy(fn, filename);
+    strncat(fn, ".json", 6);
+
+    fp = fopen(fn, "w");
+    if (fp == NULL) {
+        aws_raise_error(AWS_ERROR_OOM);
```

Comment:
Not every problem is due to OOM. There are lots of reasons `fopen` could fail. If it fails it sets thread-local `errno` [docs](https://en.cppreference.com/w/c/io/fopen)

We don't have a 1:1 mapping of every possible `errno` code to aws-error codes, but we do have a function that makes an attempt. You could use it here like:
`aws_translate_and_raise_io_error(errno)`